### PR TITLE
Create span when SamplingDecision.Drop

### LINKED ISSUE #3290: A span with a new span ID MUST be created even for a sampling decision of DROP.
# Bug Report

This bug should be present in master and at least since [0.5.0-beta](https://github.com/open-telemetry/opentelemetry-dotnet/releases/tag/0.5.0-beta) (probably since the beginning)

## Symptom

If the sampler returns a decision of DROP, StartActivity returns null.

**What is the expected behavior?**

An unsampled activity with a new span ID is created and returned (and set as current which is not spec-conformant but expected in .NET). This allows correlating logs also to unsampled operations.

Spec references:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#shouldsample
> DROP - IsRecording() == false, span will not be recorded and all events and attributes will be dropped.

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sdk-span-creation
> 3. Generate a new span ID for the Span, independently of the sampling decision. This is done so other components (such as logs or exception handling) can rely on a unique span ID, even if the Span is a non-recording instance.
> 4. [...] A non-recording span MAY be implemented using the same mechanism as when a Span is created without an SDK installed or as described in [wrapping a SpanContext in a Span](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#wrapping-a-spancontext-in-a-span).

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#wrapping-a-spancontext-in-a-span
> The behavior is defined as follows:
>
> * GetContext MUST return the wrapped SpanContext.

**What is the actual behavior?**

Null is returned for the activity, no span ID is available.

## Reproduce

There is a unit test for the wrong behavior here: https://github.com/open-telemetry/opentelemetry-dotnet/blob/641b2f76f1e3b9b1aa1697ab47ec24aed31a45c3/test/OpenTelemetry.Tests/Trace/TracerProviderSdkTest.cs#L274-L277

## Additional Context

This is a pretty big deviation from the spec that will inhibit some use cases. The point of creating a new span ID even for unsampled spans was a very deliberate decision. There were even two competing PRs https://github.com/open-telemetry/opentelemetry-specification/pull/1225 and https://github.com/open-telemetry/opentelemetry-specification/pull/1225 and it was explicitly decided on always creating a span ID.

### LINKED ISSUE #4087: OpenTelemetry.Shims.OpenTracing Invalid 'SpanContext' when not sampled
# Bug Report

List of [all OpenTelemetry NuGet packages]:

```
<PackageReference Include="OpenTelemetry" Version="1.4.0-rc.1" />
<PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.4.0-rc.1" />
<PackageReference Include="OpenTelemetry.Exporter.Jaeger" Version="1.4.0-rc.1" />
<PackageReference Include="OpenTelemetry.Exporter.Prometheus.AspNetCore" Version="1.4.0-rc.1" />
<PackageReference Include="OpenTelemetry.Extensions.Hosting" Version="1.4.0-rc.1" />
<PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" Version="1.0.0-rc9.10" />
<PackageReference Include="OpenTelemetry.Instrumentation.Hangfire" Version="1.0.0-beta.4" />
<PackageReference Include="OpenTelemetry.Instrumentation.SqlClient" Version="1.0.0-rc9.10" />
<PackageReference Include="OpenTelemetry.Shims.OpenTracing" Version="1.0.0-rc9.11" />
```

Runtime version:

* net7.0

## Symptom

When using shim package

**What is the expected behavior?**
Return some noop instance.

**What is the actual behavior?**
ArgumentException thrown from SpanShim ctor.

## Reproduce

Reproduced error:
https://github.com/konczykl/OpenTracingShimError

## Additional Context

Problem exists if we are not sampling spans - underlying activity is not created resulting in default SpanContext.
```
using (var parent = tracer.BuildSpan("parent").StartActive())
using (var child = tracer.BuildSpan("child").StartActive())
{

}
```
Parent span's activity is set as PropagationData and gets created but child's activity is null.

```
 public SpanShim(TelemetrySpan span)
{
    Guard.ThrowIfNull(span);

    if (!span.Context.IsValid) // here Context is default -> not valid
    {
        throw new ArgumentException($"Invalid '{nameof(SpanContext)}'", nameof(span.Context));
    }

    this.Span = span;
    this.spanContextShim = new SpanContextShim(this.Span.Context);
}
```

### PR DESCRIPTION
**Pull request by alanwest**
Fixes #3290 
Fixes #4087

Create spans even on a sampling decision of drop. This conforms the .NET SDK to the specification regarding [span creation](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sdk-span-creation).

> Generate a new span ID for the Span, independently of the sampling decision. This is done so other components (such as logs or exception handling) can rely on a unique span ID, even if the Span is a non-recording instance.

This may have a performance impact as an `Activity` will always be allocated irrespective of sampling decision. Notable exceptions are the ASP.NET Core and HttpClient instrumentation. The Activity created by these libraries was already always allocated.

### CODE REVIEW

**Comment by cijothomas** [comment]
Time to really comply with the spec 🤣. I vaguely remember some discussions around this in some SIG calls, and we still decided to use the current approach.... Maybe the heap allocation was a bigger concern?

**Comment by alanwest** [comment]
> Maybe the heap allocation was a bigger concern?

Yes! Was just talking to @CodeBlanch about this. Allocations gonna be painful (note the name of my branch 😆)

But this may be a hit we'll need to take... it's important for both propagating context properly and also correlation of trace data to logs and metrics.

**Comment by cijothomas** [comment]
If a root span (t1, s1) is created.. then its child gets *not* created, and then someone makes a log. the log will be parented to the (t1,s1).

With this change, the child span gets created (t1, s2, parent=s1), but wont see processor or exporter. The log will now be parented to t1,s2.. which might be an issue as log is now pointing to a parent, who is nowhere to be found....

**Comment by cijothomas** [comment]
> If a root span (t1, s1) is created.. then its child gets _not_ created, and then someone makes a log. the log will be parented to the (t1,s1).
> 
> With this change, the child span gets created (t1, s2, parent=s1), but wont see processor or exporter. The log will now be parented to t1,s2.. which might be an issue as log is now pointing to a parent, who is nowhere to be found....

unless we make log point to the span which is sampled or root. (like walk the .Parent chain until we run out of parent)...

**Comment by alanwest** [comment]
> The log will now be parented to t1,s2.. which might be an issue as log is now pointing to a parent, who is nowhere to be found....

My understanding is that this is by design. That is, incomplete traces are to be expected as well as telemetry correlated to things that happened but were not recorded.

**Comment by alanwest** [comment]
> unless we make log point to the span which is sampled or root. (like walk the .Parent chain until we run out of parent)...

If walking the chain is required, then I think this needs to be clarified by the spec.

**Comment by CodeBlanch** [comment]
A couple tags...

#3290

https://github.com/open-telemetry/opentelemetry-dotnet/pull/3828#discussion_r1007471559

**Comment by alanwest** [comment]
Closing for now... as discussed in this week's SIG, we do want to make this change, but we want to wait until after the 1.4.0 release. I will pursue fixing up the tests and open the PR back up later.

**Comment by TimothyMothra** [comment]
FYI; tested locally and confirmed this change would fix the issue reported in #4087

**Comment by cijothomas** [comment]
@TimothyMothra Thanks for investigating it more! This is tagged to resurrected for 1.5. Please chime in with a +1 to the original issue.

**Review by cijothomas — COMMENTED** [review]
LGTM. Lets add a good write in the PR description about the potential impact this has, and a changelog entry.

**Comment by cijothomas** [comment]
@Oberon00 
Could you review please?

File: src/OpenTelemetry/Trace/TracerProviderSdk.cs [hunk 1]
```diff
@@ -240,7 +239,7 @@ internal sealed class TracerProviderSdk : TracerProvider
             else if (this.sampler is AlwaysOffSampler)
             {
                 listener.Sample = (ref ActivityCreationOptions<ActivityContext> options) =>
-                    !Sdk.SuppressInstrumentation ? PropagateOrIgnoreData(options.Parent) : ActivitySamplingResult.None;
+                    !Sdk.SuppressInstrumentation ? ActivitySamplingResult.PropagationData : ActivitySamplingResult.None;
```
Comment (Oberon00) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:242]: This is extremely hard to read. There is some deep nesting of if/else/Listener callback code going on here. I would suggest to extract some (named, top-level) methods to make the code more readable/reviewable.

Comment (Oberon00) [inline src/OpenTelemetry/Trace/TracerProviderSdk.cs:242]: Am I correct that this `None` was only used as a special case for `AlwaysOffSampler`? The `ComputeActivitySamplingResult` already seems to return `ActivitySamplingResult.PropagationData` for `DROP`.

**Comment by Oberon00** [comment]
Maybe only tangentially related but:

I wonder how one would implement transparent API-only propagation in .NET, where you just echo through extracted incoming span contexts: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#behavior-of-the-api-in-the-absence-of-an-installed-sdk

**Comment by reyang** [comment]
@alanwest an Activity is fairly expensive from the garbage collection point of view. Could we explore if it's possible to defer the creation only when some code is reading it? Alternatively, maybe explore if a slim version of Activity can be used/reused rather than a full version.

**Comment by Oberon00** [comment]
>  an Activity is fairly expensive from the garbage collection point of view

Maybe System.Diagnostics.Activity is a bad match for the OpenTelemetry Span interface, since I guess the spec was written imagining that a non-recording Span can be rather efficiently implemented.

**Comment by alanwest** [comment]
> Could we explore if it's possible to defer the creation only when some code is reading it? Alternatively, maybe explore if a slim version of Activity can be used/reused rather than a full version.

I think solving this issue within the opentelemetry-dotnet project may prove challenging. Either deferring creation or implementing a slim version may require instrumentation authors to depend on OpenTelemetry.Api rather than just System.Diagnosics to retrieve current span context. I think this is something we're aiming to minimize.

> Activity is fairly expensive from the garbage collection point of view

Talking with @CodeBlanch and trying to brainstorm options we might suggest to .NET team to make Activity itself lighter weight. Though, I guess the first question is whether the size of Activity is the primary concern.

In other words, is Activity expensive from a GC point of view because of its size? Or is it expensive irrespective of size and we really just want to avoid heap allocations altogether?

**Comment by cijothomas** [comment]
> Maybe only tangentially related but:
> 
> I wonder how one would implement transparent API-only propagation in .NET, where you just echo through extracted incoming span contexts: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#behavior-of-the-api-in-the-absence-of-an-installed-sdk

This might be difficult/impossible.
One can extract ActivityContext using Propagators, but it does not have a place to put them in (so that it can be further injected to downstream calls)...unless an Activity itself is created with the extracted Context...

It might be possible to do just the baggage part, as once can use Propagators to extract it, store in Baggage.Current, and it can be injected to downstream calls as well.

**Comment by cijothomas** [comment]
> > Could we explore if it's possible to defer the creation only when some code is reading it? Alternatively, maybe explore if a slim version of Activity can be used/reused rather than a full version.
> 
> I think solving this issue within the opentelemetry-dotnet project may prove challenging. Either deferring creation or implementing a slim version may require instrumentation authors to depend on OpenTelemetry.Api rather than just System.Diagnosics to retrieve current span context. I think this is something we're aiming to minimize.
> 
> > Activity is fairly expensive from the garbage collection point of view
> 
> Talking with @CodeBlanch and trying to brainstorm options we might suggest to .NET team to make Activity itself lighter weight. Though, I guess the first question is whether the size of Activity is the primary concern.
> 
> In other words, is Activity expensive from a GC point of view because of its size? Or is it expensive irrespective of size and we really just want to avoid heap allocations altogether?

https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/Benchmarks/Trace/TraceBenchmarks.cs#L32 Its 416 B per activity minimum without any tags. Adding tags will make it ~650B/activity.

**Comment by cijothomas** [comment]
> may require instrumentation authors to depend on OpenTelemetry.Api rather than just System.Diagnosics to retrieve current span context

Something like ActivityContext.Current (like Baggage.Current) which can be thought of as just the context part, and can be valid/used even if Activity.Current is null?

**Comment by alanwest** [comment]
> Something like ActivityContext.Current (like Baggage.Current) which can be thought of as just the context part, and can be valid/used even if Activity.Current is null?

The trouble is, I'm not sure how something like this would work. Consider the following:

```csharp
public void SomeMethod()
{
    using (var span1 = source.StartActivity("SampledSpan"))
    {
        using (var span2 = source.StartActivity("NotSampledSpan"))
        {
            // span2 is null here, but let's imagine that ActivityContext.Current = span2's context
            
            // logs and exemplars are correlated to span2's context
            log.Info("Some log message.");
            histogram.Record(100);

            // span2's context is propagated
            httpClient.SendAsync(...);
        }

        // How do we know here that we are not outside the scope of span2?
        // We'd need something like ActivityContext.Current.Stop(), but how would this get invoked?
    }
}
```

**Comment by alanwest** [comment]
My sense is that we're stuck with working within the Activity API as is and optimizing perf here will require collaborating with the .NET team. Could `ActivitySource` introduce an object pool?

**Comment by alanwest** [comment]
> Its 416 B per activity minimum without any tags.

This could probably be reduced by doing something like:

```csharp
class Activity
{
   // This would be null if this Activity was not sampled
   private ActivityData Data;
   
   // This would be a valid context
   private ActivityContext context;
}

class ActivityData
{
    // Move all the fields on Activity today to this class
    // basically the fields here: https://github.com/dotnet/runtime/blob/0a15c3b97dabaf94e5d4399aef2fd48c2b3f446a/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L90-L116
}
```

But I'm not sure this helps address the primary concern. Activities would still be allocated and require garbage collection. We'd need to explore if reducing the size of Activity reduces the overhead of garbage collection.
